django/db/models/functions/window.py

```diff
@@ -0,0 +1,120 @@
+
+
+class DenseRank(Aggregate):
+    function = 'DENSE_RANK'
+    name = 'DenseRank'
+
+    def __init__(self, *expressions, **extra):
+        super().__init__(*expressions, output_field=IntegerField())
+
+
+class FirstLastValueWrapper(Aggregate):
+    def __init__(self, expression, **extra):
+        if expression is None:
+            raise ValueError('%s requires a non-null source_expression' % self.__class__.__name__)
+        super().__init__(expression, **extra)
+
+
+class FirstValue(FirstLastValueWrapper):
+    function = 'FIRST_VALUE'
+    name = 'FirstValue'
+
+
+class LagLeadFunction(Aggregate):
+    def __init__(self, expression, offset=1, default=None, **extra):
+        if expression is None:
+            raise ValueError('%s requires a non-null source expression' % self.__class__.__name__)
+
+        if offset is None or offset <= 0:
+            raise ValueError('%s requires a positive integer for the offset' % self.__class__.__name__)
+
+        if default is None:
+            super().__init__(expression, offset, **extra)
+        else:
+            super().__init__(expression, offset, default, **extra)
+
+    def _resolve_output_field(self):
+        sources = self.get_source_expressions()
+        self._output_field = sources[0].output_field
+
+
+class Lag(LagLeadFunction):
+    function = 'LAG'
+    name = 'Lag'
+
+
+class LastValue(FirstLastValueWrapper):
+    function = 'LAST_VALUE'
+    name = 'LastValue'
+
+
+class Lead(LagLeadFunction):
+    function = 'LEAD'
+    name = 'Lead'
+
+
+class NthValue(Aggregate):
+    function = 'NTH_VALUE'
+    name = 'NthValue'
+
+    def __init__(self, expression, nth=1, **extra):
+        if expression is None:
+            raise ValueError('NthValue requires a non-null source expression')
+
+        if nth is None or nth <= 0:
+            raise ValueError('NthValue requires a positive integer as for nth')
+
+        super().__init__(expression, nth, **extra)
+
+    def _resolve_output_field(self):
+        sources = self.get_source_expressions()
+        self._output_field = sources[0].output_field
+
+
+class Ntile(Aggregate):
+    function = 'NTILE'
+    name = 'Ntile'
+
+    def __init__(self, num_buckets=1, **extra):
+        if num_buckets <= 0:
+            raise ValueError('num_buckets must be greater than 0')
+        super().__init__(num_buckets, output_field=IntegerField())
+
+
+class Rank(Aggregate):
+    function = 'RANK'
+    name = 'Rank'
+
+    def __init__(self, *expressions, **extra):
+        super().__init__(*expressions, output_field=IntegerField())
+
+
+class RowNumber(Aggregate):
+    function = 'ROW_NUMBER'
+    name = 'RowNumber'
+
+    def __init__(self, *expressions, **extra):
+        super().__init__(*expressions, output_field=IntegerField())
+
+
+class PercentRank(Aggregate):
+    function = 'PERCENT_RANK'
+    name = 'PercentRank'
+
+    def __init__(self, *expressions, **extra):
+        super().__init__(*expressions, output_field=FloatField())
```

Comment:
Defining `__init__()` to specify a default `output_field` is required until we re-arrange `runtests.py` to avoid importing settings dependant modules before calling `django.setup()`.